swift/ql/lib/codeql/swift/elements/decl/EnumDecl.qll

```diff
@@ -1,4 +1,51 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.decl.EnumDecl
+private import codeql.swift.elements.decl.EnumCaseDecl
+private import codeql.swift.elements.decl.EnumElementDecl
+private import codeql.swift.elements.decl.Decl
 
-class EnumDecl extends Generated::EnumDecl { }
+/**
+ * An enumeration declaration, for example:
+ * ```
+ * enum MyColours {
+ *   case red
+ *   case green
+ *   case blue
+ * }
+ * ```
+ */
+class EnumDecl extends Generated::EnumDecl {
+  /**
+   * Gets the number of `EnumElementDecl`s in this enumeration before the `index`th member. Some
+   * of the members of an `EnumDecl` are `EnumCaseDecls` (representing the `case` lines), each of
+   * which holds one or more `EnumElementDecl`s.
+   */
+  private int countEnumElementsTo(int memberIndex) {
+    memberIndex = 0 and result = 0
+    or
+    exists(Decl prev | prev = this.getMember(memberIndex - 1) |
+      result = this.countEnumElementsTo(memberIndex - 1) + prev.(EnumCaseDecl).getNumberOfElements()
+      or
+      not prev instanceof EnumCaseDecl and
+      result = this.countEnumElementsTo(memberIndex - 1)
+    )
+  }
+
+  /**
+   * Gets the `index`th enumeration element of this enumeration (0-based).
+   */
+  final EnumElementDecl getEnumElement(int index) {
```

Comment:
I think we can simplify this not to use the `countElementsTo` helper... Something like:

```codeql

result = rank[index](int memberIndex, int elementIndex |
  this.getMember(memberIndex)
      .(EnumCaseDecl)
      .getElement(elementIndex)
  order by memberIndex, elementIndex
)

```

Reply:
You're right, and I did something similar in response to @MathiasVP s suggestion.  Your solution is slightly different, but equally correct I think (with `index` -> `index + 1` that is).

Reply:
It makes the indexes start a 0 instead of 1.  I believe this is because `rank` indexes start at 1.

Reply:
Yes. `rank` is 1-based.